osbenchmark/worker_coordinator/worker_coordinator.py

```diff
@@ -198,6 +203,223 @@ def __init__(self, metrics, next_task_scheduled_in):
+        elif errors:
+            self.logger.info("Error messages detected, scaling down...")
+            self.state = FeedbackState.SCALING_DOWN
+            with self.queue_lock:  # Block producers while scaling down.
+                self.scale_down()
+            self.logger.info("Clients scaled down. Active clients: %d", self.total_active_client_count)
+            self.last_error_time = current_time
+        elif self.state == FeedbackState.NEUTRAL:
+            self.max_stable_clients = max(self.max_stable_clients, self.total_active_client_count) # update the max number of stable clients
+            if (current_time - self.last_error_time >= self.POST_SCALEDOWN_SECONDS and
+                current_time - self.last_scaleup_time >= self.WAKEUP_INTERVAL):
+                self.logger.info("No errors in the last %d seconds, scaling up", self.POST_SCALEDOWN_SECONDS)
+                self.state = FeedbackState.SCALING_UP
+
+        if self.state == FeedbackState.SCALING_UP:
+            self.logger.info("Scaling up...")
+            self.scale_up()
+            self.logger.info("Clients scaled up. Active clients: %d", self.total_active_client_count)
+            self.state = FeedbackState.NEUTRAL
+
+    def scale_down(self, scale_down_percentage=0.10) -> None:
+        try:
+            clients_to_pause = int(self.total_active_client_count * scale_down_percentage)
+            if clients_to_pause <= 0:
+                self.logger.info("No clients to pause during scale down")
+                return
+
+            # Create a flattened list of (worker_id, client_id) tuples for all active clients
+            all_active_clients = []
+            for worker_id, client_states in self.shared_client_states.items():
+                for client_id, status in client_states.items():
+                    if status:  # Only include active clients
+                        all_active_clients.append((worker_id, client_id))
+
+            # If we need to pause more clients than are active, adjust the count
+            clients_to_pause = min(clients_to_pause, len(all_active_clients))
+
+            # Select clients to pause - randomly sample for better distribution
+            clients_to_pause_indices = random.sample(range(len(all_active_clients)), clients_to_pause)
+            clients_to_pause_list = [all_active_clients[i] for i in clients_to_pause_indices]
+
+            # Pause the selected clients in a single pass
+            for worker_id, client_id in clients_to_pause_list:
+                self.shared_client_states[worker_id][client_id] = False
+                self.total_active_client_count -= 1
+
+            self.logger.info("Scaling down complete. Paused %d clients", clients_to_pause)
+        finally:
+            self.state = FeedbackState.SLEEP
+            self.clear_queue()
+            self.sleep_start_time = self.last_scaleup_time = time.perf_counter()
+
+    def scale_up(self, n_clients=5) -> None:
```

Comment:
Super nit: We can turn this into a class variable so that there's a central location to update the default value in the future